internal/retry/retry.go

```diff
@@ -14,44 +15,66 @@ import (
 // IntervalRetrier retries a call with an interval. The call is defined in the Doer property.
 type IntervalRetrier struct {
 	interval  time.Duration
+	deadline  time.Time
 	doer      Doer
 	clock     clock.WithTicker
 	retriable func(error) bool
 }
 
 // NewIntervalRetrier returns a new IntervalRetrier. The optional clock is used for testing.
-func NewIntervalRetrier(doer Doer, interval time.Duration, retriable func(error) bool, optClock ...clock.WithTicker) *IntervalRetrier {
+func NewIntervalRetrier(doer Doer, interval time.Duration, deadline time.Time, retriable func(error) bool, optClock ...clock.WithTicker) *IntervalRetrier {
 	var clock clock.WithTicker = clock.RealClock{}
 	if len(optClock) > 0 {
 		clock = optClock[0]
 	}
 
 	return &IntervalRetrier{
 		interval:  interval,
+		deadline:  deadline,
 		doer:      doer,
 		clock:     clock,
 		retriable: retriable,
 	}
 }
 
-// Do retries performing a call until it succeeds, returns a permanent error or the context is cancelled.
+// Do retries performing a call until it succeeds, returns a permanent error, the deadline passes or the context expires.
+//
+// The doer will be called at least once, regardless of context or deadline expiration.
+// The return value is either nil if doer was successful, or the last error returned by the doer
+// wrapped in an error describing the cause of retry failure (non-retriable, context expired,
+// deadline passed).
 func (r *IntervalRetrier) Do(ctx context.Context) error {
 	ticker := r.clock.NewTicker(r.interval)
 	defer ticker.Stop()
 
+	timer := r.clock.NewTimer(r.deadline.Sub(r.clock.Now()))
+	defer timer.Stop()
+
 	for {
 		err := r.doer.Do(ctx)
```

Comment:
> If the retry loop expiration is controlled by the input context, and
that context is also passed to the wrapped call, it is very likely that
we will see the context expiration error as last inner error. To avoid
this, we don't attach a deadline to the outer context, but configure the
retrier with an explicit deadline. 


The problem is that when this ctx times out, we get a unintesting error, right? 
Would it be simpler/also feasible to simply save the lastErr in the IntervalRetrier and when we Do returns an error, we check if the context is done and then return both ctx.Err and r.lastErr (or we just always return both)?

Reply:
That's what I'm doing below, isn't it?

```go
		case <-timer.C():
			return fmt.Errorf("retry deadline exceeded: last error: %w", err)
		case <-ctx.Done():
			return fmt.Errorf("%w: last error: %w", ctx.Err(), err)
```

Just from the returned error of the doer, I can't tell whether this was caused by the deadline externally imposed on me, the IntervalRetrier, or by a more strict deadline internal to the doer. Also, we'd have to track the last two errors, because the last error might just be the context error that I'm already aware of.

Reply:
My idea is to not have another timeout/ctx I want to think about. The context I pass to `retrier.Do` should be the context of the retrier itself. If I want the retrier to end after a certain amount I pass a `context.WithTimeout()`. 

I think the following things always return the interesting error:

1. Do returns err
2. if err is not retriable, return the error and the lastErr (this will become important later)
3. if err is retriable, check if ctx is done, if so return err and lastErr
4. save err as lastErr

if the inner error is a deadline exceeded because the retrier ctx expired then we always add the error before this deadline exceeded.

Reply:
> I'm not entirely happy with the most recent error being replaced by the last error on context expiration, 

You mean because if the .Do returns a context deadline exceeded, the error is retriable but in most cases uninsteresting?
Maybe we can also write a error filter function that we give to the retrier like we do with the retriable(err error) function.

Also I think that using only the context is an abstraction that makes all the invocations of the retrier easier, since on the invocation site, the calling function might themselves not know if the context has a timeout.

Reply:
Sorry, there may have been a misunderstanding: I was arguing in favour of your suggestion, using only the context for controlling the retry loop.